app/controllers/projects_controller.rb

```diff
@@ -65,6 +61,14 @@ def visuals
     end
   end
 
+  def favorite
+    if current_user.favorites.include? @project
```

Comment:
this is a bit dangerous. if the call is sent twice advertently, it will undo what it is supposed to do.  can you do add_favorite and remove_favorite instead?

Reply:
yes @yunfangjuan, this method is meant to do exactly that, toggle favorite/unfavorite, currently on the client side, the user will only be allowed to make the call for unfavorite if the project is favorited already and vice versa.

Reply:
well but I think from the backend standpoint, it's not a deterministic action. sometimes, it's delete. sometimes it's add. It can make the future debugging hard to do. I think we should still break it do add_favorite and remove_favorite. if the frontend accidentally call the api twice (which could happen in some brower), it won't do anything.